internal/pkg/levee/testdata/src/levee_analysistest/example/suppression/tests.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package suppression contains tests for false positive suppression.
+package suppression
+
+import (
+	"levee_analysistest/example/core"
+)
+
+func TestNotSuppressed(s core.Source) {
+	core.Sink(s) // want "a source has reached a sink"
+}
+
+func TestOnelineLineComment(s core.Source) {
+	// levee.DoNotReport
+	core.Sink(s)
+}
+
+func TestOnelineGeneralComment(s core.Source) {
+	/* levee.DoNotReport */
+	core.Sink(s)
+}
+
+func TestInlineLineComment(s core.Source) {
+	core.Sink(s) // levee.DoNotReport
+}
+
+func TestInlineGeneralComment(s core.Source) {
+	core.Sink(s) /* levee.DoNotReport */
+}
+
+func TestMultilineLineComment(s core.Source) {
+	// Line 1
+	// levee.DoNotReport
+	// Line 3
+	core.Sink(s)
+}
+
+func TestMultilineGeneralComment(s core.Source) {
+	/*
+		Line 1
+		levee.DoNotReport
+		Line 3
+	*/
+	core.Sink(s)
+}
+
+func TestAdjacentReports(s core.Source) {
+	core.Sink(s) // levee.DoNotReport
+	core.Sink(s) // want "a source has reached a sink"
+}
+
+func TestReportsSeparatedByLineComment(s core.Source) {
+	core.Sink(s) // want "a source has reached a sink"
+	// levee.DoNotReport
+	core.Sink(s)
+}
+
+func TestReportsSeparatedByGeneralComment(s core.Source) {
+	core.Sink(s) /*
+		levee.DoNotReport
+	*/
+	core.Sink(s) // want "a source has reached a sink"
+}
+
+func TestLineCommentBeforeGeneralComment(s core.Source) {
```

Comment:
Good test cases, thanks! I did some more digging around this and I ran into some weird things, e.g.:

```go
	core.Sink( // levee.DoNotReport
		"arg 1",
		s,
		"arg 3")

```

What node is the comment associated with in this example? That's right! The `Sink` identifier! Which makes sense, but as a user I would expect this to suppress the report.

It turns out it's straightforward to support that case with the information that we have in the `ast.CallExpr` (first node returned by `PathEnclosingInterval`).

We could support suppression on arguments as well, but I feel like that might be misleading, e.g. a user might think "here, let me suppress *just* this specific argument" and not realize that it suppresses the entire call. Correctly handling suppression in `Args` requires inspecting the entire tree below the `ast.CallExpr`. Indeed, consider this case:

```go
core.Sink(
  []interface{}{
    s, // levee.DoNotReport
  })
```

(In the example above, it's not sufficient to just look at the nodes in `ast.CallExpr.Args`, and in the general case we really need to dig into the tree.)

So we could definitely suppress the entire call when the suppression is associated with a specific argument, but that feels misleading to me. The following case is a bit unfortunate, though:

```go
core.Sink(s,  // levee.DoNotReport
  "foo", "bar")
```

In the above case, the comment is associated with `s`, not with the call, so currently this doesn't suppress the report. (But some might expect it to.)

I think as long as this is clearly documented, this is reasonable behavior. WDYT?

Reply:
> I think as long as this is clearly documented, this is reasonable behavior. WDYT?

I think that's one of the two guiding stars.

I'd say the other one is to start simple.  I'm perfectly happy to have a more restrictive use pattern and, should we get requests for finer-grained control, iterate then.

Here I'm asking because I'm still not familiar with the `ast` graph, but how would it look if we looked at all the ancestry and descendants at a given sink call?  (Would that even be possible without a huge rework?)  To my crude understanding, we'd end up with three cases:
* The comment is on an descendant argument, so we exclude the call entirely (and document that it doesn't matter which arg it's on)
* The comment is on the call itself, so is naturally excluded
* The comment is on an ancestor, implying broader exclusion.

That last bullet makes me think we could get from a comment the sort of exclusion possible in configuration now - a function-level comment would exclude all contained calls, and possible a package-level comment could exclude everything there.  That's probably opening a granularity can of worms, so I could be absolutely fine not going that route.

It also brings into question nested calls.  What happens in the following case?

```golang
func SinkAndReturnError(s core.Source) error {  // <- This is declared as a sink
  return fmt.Errorf("%v", s)
}

func notASink(interface{}) {}

func TestInnerSink(s core.Source) {
  notASink(
    SinkAndReturnError(s)  // levee.DoNotReport
  )
}
```

Ignoring the taint propagation aspect, that might be a case we want to cover.  Also reasonable as a candidate for iteration, but something we should document one way or t'other.

Reply:
We definitely can handle both args (visit the tree below the call) and ancestors (use `astutil.PathEnclosingInterval`), but for now in my opinion we don't want either of those.

Concerning nested calls, good point. The results are interesting.

This works:

```go
	fmt.Println(
		// levee.DoNotReport
		core.SinkAndReturn(s),
	)
```

This doesn't work:

```
	fmt.Println(
		core.SinkAndReturn(s), // levee.DoNotReport
	)
```

In the above, the comment is associated, rather suprisingly, with the `s` identifier (`s` is a `core.Source` value).

Reply:
> For now, I think it would be reasonable to open an issue for that second case, and include it in the tests with an appropriate TODO referencing the issue. WDYT?

Iterating on this sounds good to me.

Reply:
Good. I've added an issue.